roachprod: stage arm64 binary

srosenberg opened the pull request:
Add `--arch` to override binary's architecture and refactor.
As of this change, `roachprod stage` is able to stage both
amd64 and arm64 on linux and darwin.
In conjunction with the previous change [1], roachprod
now uses arm64-based AMI for graviton2/graviton3 machines.

Below is an example of how to create a VM with graviton3,
```
roachprod create -n1 --clouds aws --aws-machine-type m7g.2xlarge --local-ssd=false $CRL_USERNAME-test
roachprod stage --arch arm64 $CRL_USERNAME-test release v23.1.0-rc.2
roachprod start $CRL_USERNAME-test
```

[1] https://github.com/cockroachdb/cockroach/pull/103236

Epic: none
Release note: None

srosenberg committed:
roachprod: stage arm64 and FIPS-enabled binaries

Add `--arch` to override binary's architecture and refactor.
As of this change, `roachprod stage` is able to stage both
amd64 and arm64 on linux and darwin, as well as FIPS-enabled
binary built for amd64.
In conjunction with the previous change [1], roachprod
now uses arm64-based AMI for graviton2/graviton3 machines.

Below is an example of how to create a VM with graviton3,
```
roachprod create -n1 --clouds aws --aws-machine-type m7g.2xlarge --local-ssd=false $CRL_USERNAME-test
roachprod stage --arch arm64 $CRL_USERNAME-test release v23.1.0-rc.2
roachprod start $CRL_USERNAME-test
```

[1] https://github.com/cockroachdb/cockroach/pull/103236

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103243)
<!-- Reviewable:end -->

rail reviewed (approved):
Just one question inline.

srosenberg commented on pkg/cmd/roachprod/flags.go:
```diff
@@ -217,10 +218,9 @@ Default is "RECURRING '*/15 * * * *' FULL BACKUP '@hourly' WITH SCHEDULE OPTIONS
+	stageCmd.Flags().StringVar(&stageArch, "arch", "", "architecture override for staged binaries")
 	stageCmd.Flags().StringVar(&stageDir, "dir", "", "destination for staged binaries")
 
-	stageURLCmd.Flags().StringVar(&stageOS, "os", "", "operating system override for staged binaries")
```
Good catch! I thought this was a duplicate of `stageCmd` :) didn't realize we had a separate command to print the URL. I'll resolve.

herkolategan reviewed (approved):
> Reviewed 9 of 10 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail and @smg260)



<!-- Sent from Reviewable.io -->

smg260 reviewed (commented):
> Reviewed 1 of 7 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail and @srosenberg)

___
*[`pkg/roachprod/roachprod.go` line 519 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeOChZEKY6W8TGoeyZ:-NVeOChZEKY6W8TGoey_:b-nyxphk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/roachprod.go#L519)):*
> ```Go
> 	os := "linux"
> 	arch := "amd64"
> ```

this logic looks a little clunky - is it deliberate?
`isLocal()` as the first condition seems cleaner, with parameters always taking precedence

```
os := ...
arch := ...

if c.IsLocal() {

}

if stageOs != "" { .. override }
if stageArch != "" {.. override }
```


<!-- Sent from Reviewable.io -->

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail and @smg260)

___
*[`pkg/roachprod/roachprod.go` line 519 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeOChZEKY6W8TGoeyZ:-NVeS2AY0Mt1_cuLGaCH:bpll2ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/roachprod.go#L519)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

this logic looks a little clunky - is it deliberate?
`isLocal()` as the first condition seems cleaner, with parameters always taking precedence

```
os := ...
arch := ...

if c.IsLocal() {

}

if stageOs != "" { .. override }
if stageArch != "" {.. override }
```
</blockquote></details>

Good catch! Technically, you should be allowed to stage an arch that's different from your local; e.g., running (emulated) amd64 on apple silicon. Either way, `IsLocal` should logically come first, and I'll add a warning if someone is attempting to stage an os/arch which differs from local.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail, @smg260, and @srosenberg)

___
*[`pkg/roachprod/roachprod.go` line 519 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeOChZEKY6W8TGoeyZ:-NVepqd20sHhLPOEzW8X:b-lw0jb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/roachprod.go#L519)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Good catch! Technically, you should be allowed to stage an arch that's different from your local; e.g., running (emulated) amd64 on apple silicon. Either way, `IsLocal` should logically come first, and I'll add a warning if someone is attempting to stage an os/arch which differs from local.
</blockquote></details>

Could these defaults also be coming from the command line? (i.e., `StringVar(&stageArch, "arch", defaultArch, ...)` where `defaultArch = "amd64"`? Then: 1) we don't need to check for command line overwrites; 2) `--help` is more explicit about the default values used; and 3) we don't need to support empty `stageArch` values (see comment on `archInfoForOS`).
___
*[`pkg/roachprod/install/staging.go` line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeqLZs9aFIgvbpsQGl:-NVeqLZt3zLAFfMk3fcj:bddmhll) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/install/staging.go#L98)):*
> ```Go
> 			return darwin_arm64_ArchInfo, nil
> 		}
> 		return darwin_x86_64_ArchInfo, nil
> ```

One less desirable property of this approach is that a typo (or even a genuinely unsupported arch) will silently use `x86_64` (e.g., `--arch amd63`)


<!-- Sent from Reviewable.io -->

srosenberg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rail, @renatolabs, and @smg260)

___
*[`pkg/roachprod/roachprod.go` line 519 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeOChZEKY6W8TGoeyZ:-NVgtdCUE3-ysOdW8d1p:b-r58gal) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/roachprod.go#L519)):*
> Could these defaults also be coming from the command line?

Sorta. The multi-value validation story for cobra is not great. Technically, you can define a custom `pflag.Value` interface, but that seems like an overkill. Instead, I borrowed the approach from the `roachtest` CLI which installs a custom validation function by using the`PersistentPreRun` hook.

That takes care of CLI validation, but recall that roachprod is also used as an API. `archInfoForOS` is used internally for staging, so added a validation bit there too. It's still not bullet-proof but maybe good enough, considering we currently don't validate a bunch of other args; i.e., to be revisited.
___
*[`pkg/roachprod/install/staging.go` line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-NVeqLZs9aFIgvbpsQGl:-NVgvj3e8tu0CHPzBLla:b-o9rn0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f6ab465fa5285a954e20da10ec94850c5fd9ee6/pkg/roachprod/install/staging.go#L98)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

One less desirable property of this approach is that a typo (or even a genuinely unsupported arch) will silently use `x86_64` (e.g., `--arch amd63`)
</blockquote></details>

Yep, added explicit precondition to validate supported values. I am not foreseeing other architectures we'd support in the near future, but you never know. If a new one does pop up, then we can refactor it to `enum` and use a custom `pflag.Value` (with shell autocompletion). For now, that seems like an overkill, would you agree?


<!-- Sent from Reviewable.io -->

smg260 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 7 files at r2, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103243#-:-NVkThe3DwYi2nHNf1Wp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rail and @renatolabs)



<!-- Sent from Reviewable.io -->

srosenberg commented:
After a final check, I discovered that the binary staging logic drifted from its source of truth (`pkg/release`). I've updated and tested all `--os` and `--arch` combinations. Additionally, I've added `fips` under `--arch`; it's a leaky abstraction, but ok for now. We'll clean it up in subsequent iterations.

PTAL!

srosenberg commented:
TFTR! Next PR adds a few small improvements, so merging this one since it seems functionally correct.

bors r=rail,herkolategan,smg260

srosenberg commented:
bors retry
